wit/types.wit

```diff
@@ -483,88 +400,15 @@ interface types {
-  /// Request, or sending a late status code for the Response.
-  resource outgoing-body {
-
-    /// Returns a stream for writing the body contents.
-    ///
-    /// The returned `output-stream` is a child resource: it must be dropped
-    /// before the parent `outgoing-body` resource is dropped (or finished),
-    /// otherwise the `outgoing-body` drop or `finish` will trap.
-    ///
-    /// Returns success on the first call: the `output-stream` resource for
-    /// this `outgoing-body` may be retrieved at most once. Subsequent calls
-    /// will return error.
-    write: func() -> result<output-stream>;
-
-    /// Finalize an outgoing body, optionally providing trailers. This must be
-    /// called to signal that the response is complete. If the `outgoing-body`
-    /// is dropped without calling `outgoing-body.finalize`, the implementation
-    /// should treat the body as corrupted.
-    ///
-    /// Fails if the body's `outgoing-request` or `outgoing-response` was
-    /// constructed with a Content-Length header, and the contents written
-    /// to the body (via `write`) does not match the value given in the
-    /// Content-Length.
-    finish: static func(
-      this: outgoing-body,
-      trailers: option<trailers>
-    ) -> result<_, error-code>;
-  }
-
-  /// Represents a future which may eventaully return an incoming HTTP
-  /// Response, or an error.
-  ///
-  /// This resource is returned by the `wasi:http/outgoing-handler` interface to
-  /// provide the HTTP Response corresponding to the sent Request.
-  resource future-incoming-response {
-    /// Returns a pollable which becomes ready when either the Response has
-    /// been received, or an error has occured. When this pollable is ready,
-    /// the `get` method will return `some`.
-    subscribe: func() -> pollable;
-
-    /// Returns the incoming HTTP Response, or an error, once one is ready.
-    ///
-    /// The outer `option` represents future readiness. Users can wait on this
-    /// `option` to become `some` using the `subscribe` method.
+    /// Returns the resource corresponding to the Body for this Response.
     ///
-    /// The outer `result` is used to retrieve the response or error at most
-    /// once. It will be success on the first call in which the outer option
-    /// is `some`, and error on subsequent calls.
-    ///
-    /// The inner `result` represents that either the incoming HTTP Response
-    /// status and headers have recieved successfully, or that an error
-    /// occured. Errors may also occur while consuming the response body,
-    /// but those will be reported by the `incoming-body` and its
-    /// `output-stream` child.
-    get: func() -> option<result<result<incoming-response, error-code>>>;
-
+    /// Returns success on the first call: the `body` resource for this
+    /// `response` can be retrieved at most once. Subsequent calls will return
+    /// error.
+    body: func() -> result<body>;
```

Comment:
I wonder if, for the same reason that we have a `static finish` that consumes the body to produce the trailers (thereby avoiding leaving `body` resources in an indeterminate state), if we should have a `static consume` on `request`/`response` that consumes the `request`/`response` (thereby avoiding the "exists, but doesn't have a valid body, so probably don't hand off this request/response to anyone" state).  This would be complemented by a `take-headers` method that lets you hold onto the headers even after consuming the request/response.

Once you make that change in your head, I think we can simplify one step further and have, directly on both `request` and `response`:
* `consume: func() -> result<stream<u8>>`
* `static finish: func(request|response) -> future<trailers>`

Lastly, because with streams it is clear exactly how many elements are successfully passed from producer to consumer, we could even remove the "can only call `consume` once rule" and say that if you call `consume` and drop the stream, you can call `consume` again and pick up where you left off (repeatedly up until you call `finish`).  There would still need to be a `result` to fail if you call `consume` concurrent with a previous call to `consume`.  I think this might even match the semantics of Web Streams, which have a [locked](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/locked) state that is set when you create a stream "reader" and can be unlocked by dropping that reader (and thus our `stream` handle that can be dropped by the caller is rather like the Web Stream reader).  Overall, what this would allow is for component C1 to be given a request, to consume some bytes, and then pass on the request to C2, which consumes some more.

WDYT?

Reply:
SGTM!

Reply:
I just pushed an update with those changes.